File: benchmarks/java/RegexRedux.java
/*
   The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Francois Green
*/

package benchmarks.java;

import benchmarkhandle.*;
import benchmarks.BaseBenchmark;

import java.io.*;

import java.util.*;
import java.util.Map.Entry;
import java.util.function.*;
import java.util.regex.*;

import static java.util.stream.Collectors.*;

public class RegexRedux extends BaseBenchmark {

    public static void main(String[] args) throws IOException {
        RegexRedux regexRedux=new RegexRedux();
        BenchmarkHandler benchmarkHandler=new BenchmarkHandler();
        BenchmarkMessage benchmarkMessage = new BenchmarkMessage(BenchmarkType.BINARYTREE, BenchmarkImplementation.JAVA, BenchmarkMetric.GARBAGE_COLLECTION);
        benchmarkHandler.startMeasuringMetrics(benchmarkMessage,regexRedux,args);
    }

    @Override
    public void initAlgorithm(String[] args) throws Exception {
        InputStream stream = new FileInputStream("/Users/enes/Documents/intelljprojects/src/regex.txt");

        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        {
            byte[] buf = new byte[65536];
            int count;
            while ((count = stream.read(buf)) > 0) {
                baos.write(buf, 0, 10245);

            }
        }
        final String input = baos.toString("US-ASCII");

        final int initialLength = input.length();

        final String sequence = input.replaceAll(">.*\n|\n", "");

        final int codeLength = sequence.length();

        final List<String> variants = Arrays.asList("agggtaaa|tttaccct",
                "[cgt]gggtaaa|tttaccc[acg]",
                "a[act]ggtaaa|tttacc[agt]t",
                "ag[act]gtaaa|tttac[agt]ct",
                "agg[act]taaa|ttta[agt]cct",
                "aggg[acg]aaa|ttt[cgt]ccct",
                "agggt[cgt]aa|tt[acg]accct",
                "agggta[cgt]a|t[acg]taccct",
                "agggtaa[cgt]|[acg]ttaccct");

        BiFunction<String, String, Entry<String, Long>> counts = (v, s) -> {
            //Java 9 Matcher.results isn't off by one
            Long count = Pattern.compile(v).splitAsStream(s).count() - 1;
            return new AbstractMap.SimpleEntry<>(v, count);
        };

        final Map<String, Long> results = variants.parallelStream()
                .map(variant -> counts.apply(variant, sequence))
                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));

        variants.forEach(variant -> System.out.println(variant + " " + results.get(variant)));

        final Map<String, String> replacements = new LinkedHashMap<>();//Only works with LinkedHashMap
        {
            replacements.put("tHa[Nt]", "<4>");
            replacements.put("aND|caN|Ha[DS]|WaS", "<3>");
            replacements.put("a[NSt]|BY", "<2>");
            replacements.put("<[^>]*>", "|");
            replacements.put("\\|[^|][^|]*\\|", "-");
        }

        String buf = sequence;
        for (Map.Entry<String, String> entry : replacements.entrySet()) {
            buf = Pattern.compile(entry.getKey()).matcher(buf).replaceAll(entry.getValue());
        }

        System.out.println();
        System.out.println(initialLength);
        System.out.println(codeLength);
        System.out.println(buf.length());
    }
}

File: benchmarks/kotlinconverted/RegexRedux.kt
package benchmarks.kotlinconverted

/*
   The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Francois Green
*/

import benchmarkhandle.*
import benchmarks.BaseBenchmark
import java.io.*

import java.util.*
import kotlin.collections.Map.Entry
import java.util.function.*
import java.util.regex.*

import java.util.stream.Collectors.*

object RegexRedux : BaseBenchmark() {

    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val regexRedux=RegexRedux
        val benchmarkHandler=BenchmarkHandler()
        val benchmarkMessage= BenchmarkMessage(BenchmarkType.FANNKUCH_REDUX, BenchmarkImplementation.KOTLIN_CONVERTED,BenchmarkMetric.MEMORY_CONSUMPTION)

          benchmarkHandler.startMeasuringMetrics(benchmarkMessage,regexRedux,args)
    }

    override fun initAlgorithm(args: Array<out String>?) {

        val stream = FileInputStream("/Users/enes/Documents/intelljprojects/src/regex.txt")

        val baos = ByteArrayOutputStream()
        run {
            val buf = ByteArray(65536)
            var count: Int
            while ((stream.read(buf)) > 0) {
                baos.write(buf, 0, 10245)

            }
        }
        val input = baos.toString("US-ASCII")

        val initialLength = input.length

        val sequence = input.replace(">.*\n|\n".toRegex(), "")

        val codeLength = sequence.length

        val variants = Arrays.asList(
            "agggtaaa|tttaccct",
            "[cgt]gggtaaa|tttaccc[acg]",
            "a[act]ggtaaa|tttacc[agt]t",
            "ag[act]gtaaa|tttac[agt]ct",
            "agg[act]taaa|ttta[agt]cct",
            "aggg[acg]aaa|ttt[cgt]ccct",
            "agggt[cgt]aa|tt[acg]accct",
            "agggta[cgt]a|t[acg]taccct",
            "agggtaa[cgt]|[acg]ttaccct"
        )

        fun patternEntry(v: String, s: String): AbstractMap.SimpleEntry<String, Long> {
            val count = Pattern.compile(v).splitAsStream(s).count() - 1
            return AbstractMap.SimpleEntry<String, Long>(v, count)
        }

        val results = variants.parallelStream()
            .map { variant -> patternEntry(variant, sequence) }
            .collect(toMap({ return@toMap it.key }, AbstractMap.SimpleEntry<String, Long>::value))

        variants.forEach { variant -> println(variant + " " + results[variant]) }

        val replacements = LinkedHashMap<String, String>()//Only works with LinkedHashMap
        run {
            replacements["tHa[Nt]"] = "<4>"
            replacements["aND|caN|Ha[DS]|WaS"] = "<3>"
            replacements["a[NSt]|BY"] = "<2>"
            replacements["<[^>]*>"] = "|"
            replacements.put("\\|[^|][^|]*\\|", "-")
        }

        var buf = sequence
        for ((key, value) in replacements) {
            buf = Pattern.compile(key).matcher(buf).replaceAll(value)
        }

        println()
        println(initialLength)
        println(codeLength)
        println(buf.length)
    }
}
